primitives/src/persistent.rs

```diff
@@ -0,0 +1,240 @@
+                let result = (0..len).map(|_| Decode::decode(input)).collect();
+                input.ascend_ref();
+                result
+            })
+        }
+    }
+
+    impl<T> FromIterator<T> for PVector<T> {
+        fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+            Self(rpds::VectorSync::from_iter(iter))
+        }
+    }
+
+    impl<T: TypeId> TypeId for PVector<T> {
+        fn id() -> String {
+            format!("Vec<{}>", T::id())
+        }
+    }
+
+    impl<T: IntoSchema> IntoSchema for PVector<T> {
+        fn type_name() -> String {
+            format!("Vec<{}>", T::type_name())
+        }
+        fn update_schema_map(map: &mut iroha_schema::MetaMap) {
+            if !map.contains_key::<Self>() {
+                map.insert::<Self>(iroha_schema::Metadata::Vec(iroha_schema::VecMeta {
+                    ty: core::any::TypeId::of::<T>(),
+                }));
+
+                T::update_schema_map(map);
+            }
+        }
+    }
+}
+
+/// Persistent hash set
+pub mod set {
+    use super::*;
+
+    /// Persistent set
+    #[derive(Serialize, Deserialize, Deref, DerefMut, Clone, Debug)]
+    pub struct PHashTrieSet<T: Eq + Hash>(rpds::HashTrieSetSync<T>);
+
+    impl<T: Eq + Hash> PHashTrieSet<T> {
+        /// Construct [`Self`]
+        pub fn new() -> Self {
+            Self(rpds::HashTrieSet::new_sync())
+        }
+    }
+
+    impl<T: Eq + Hash> Default for PHashTrieSet<T> {
+        fn default() -> Self {
+            Self(rpds::HashTrieSetSync::default())
+        }
+    }
+
+    impl<T: Encode + Eq + Hash> Encode for PHashTrieSet<T> {
+        #[allow(clippy::arithmetic_side_effects)]
+        fn size_hint(&self) -> usize {
+            mem::size_of::<u32>() + mem::size_of::<T>() * self.size()
+        }
+
+        fn encode_to<W: Output + ?Sized>(&self, dest: &mut W) {
+            let len = self.size();
+            Compact(u32::try_from(len).expect("Length is above u32::max")).encode_to(dest);
+
+            for element in self.iter() {
+                element.encode_to(dest);
+            }
+        }
+    }
+
+    impl<T: Decode + Eq + Hash> Decode for PHashTrieSet<T> {
+        fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+            <Compact<u32>>::decode(input).and_then(move |Compact(len)| {
+                input.descend_ref()?;
+                let result = (0..len).map(|_| Decode::decode(input)).collect();
+                input.ascend_ref();
+                result
+            })
+        }
+    }
+
+    impl<T: Eq + Hash> FromIterator<T> for PHashTrieSet<T> {
+        fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+            Self(rpds::HashTrieSetSync::from_iter(iter))
+        }
+    }
+
+    impl<T: TypeId + Eq + Hash> TypeId for PHashTrieSet<T> {
+        fn id() -> String {
+            format!("Vec<{}>", T::id())
+        }
+    }
+
+    impl<T: IntoSchema + Eq + Hash> IntoSchema for PHashTrieSet<T> {
```

Comment:
Looks like we don't allow *Hash-based* structures in our schema. Are you sure we need this impl?

Reply:
And what's the problem with that?